src/index.ts

```diff
@@ -83,695 +115,648 @@ if (!Object.keys) {
-        compressArg(`def`, args);
-        compressArg(`ext`, args);
-        compressArg(`qual`, args);
-        compressArg(`size`, args);
-      }
+      file.write(`\n`);
     }
-  }
 
-  const docName: string = app.activeDocument.name.replace(/\.[a-z]{3,4}$/i, ``);
-  const docPath: Folder = app.activeDocument.path;
-  const file: File = File(`${docPath}/${docName}-assets/responsive-image.html`);
-
-  if (file.exists) {
-    file.remove();
+    file.writeln(`${tabs}  "`);
   }
 
-  file.encoding = 'utf-8';
-  file.open('w');
-
-  const tab: string = `  `;
-
-  if (hasMultipleContexts(sortedContexts)) {
-    file.writeln(`<picture>`);
-
-    for (let i: number = sortedContexts.length; i > 0; i--) {
-      const sortedContext: TShirtSize = sortedContexts[i - 1];
-      const assets: Assets = img.contexts[sortedContext]!.assets;
-
-      if (i === 1) {
-        file.writeln(`${tab}<img`);
+  file.writeln(`${tabs}/>`);
+};
 
-        if (hasAlt(img.alt)) {
-          file.writeln(`${tab}${tab}alt="${img.alt}"`);
+/**
+ * Generates a picture tag from `resImg`, then writes it to `file`.
+ *
+ * @since 1.0.0
+ *
+ * @param {Object} file The file to write the img tag to.
+ * @param {Object} resImg The object containing alt text, src directory path and
+ * context max-widths.
+ * @returns {void} This function doesn't have a return statement.
+ */
+const generatePictureTag = (file: any, resImg: ResponsiveImage) => {
+  file.writeln(`<picture>`);
+
+  const contexts: AssetContext[] = sortContexts(Object.keys(assets));
+  const maxWidths: AssetContext[] = sortContexts(Object.keys(resImg.maxWidths));
+
+  for (let i: number = contexts.length; i > 0; i--) {
+    const contextualAssets: Asset[] = sortAssets(assets[contexts[i - 1]]!);
+    const minWidth: number = resImg.maxWidths[maxWidths[i - 2]]! + 1;
+
+    if (i > 1) {
+      file.writeln(`  <source`);
+      file.writeln(`    media="(min-width: ${minWidth / 16}em)"`);
+      file.writeln(`    srcset="`);
+
+      for (
+        let i: number = 0, len: number = contextualAssets.length;
+        i < len;
+        i++
+      ) {
+        file.write(`      ${resImg.srcDir}/${contextualAssets[i].fileName}`);
+
+        if (contextualAssets[i].def > 1) {
+          file.write(` ${contextualAssets[i].def}x`);
         }
 
-        for (let i: number = 0, len: number = assets.length; i < len; i++) {
-          const asset: Asset = assets[i];
-
-          if (i === 0) {
-            file.writeln(`${tab}${tab}src="${getAssetSrc(asset)}"`);
-
-            if (len > 1) {
-              file.writeln(`${tab}${tab}srcset="`);
-            }
-          } else {
-            if (i === len - 1) {
-              file.writeln(`${tab}${tab}${tab}${getAssetSrc(asset)}`);
-              file.writeln(`${tab}${tab}"`);
-            } else {
-              file.writeln(`${tab}${tab}${tab}${getAssetSrc(asset)},`);
-            }
-          }
+        if (i < len - 1) {
+          file.write(`,`);
```

Comment:
how about extracting (a) function(s) to generalise this <source> element creation? It will simplify the code and probz even improve the efficiency

Reply:
I guess what I mean is 
if i > 0 generateSourceTag() else generateImgTag()

plus separate functions for the inner stuff when generating both (remove all code duplication cancer): when generating each of these "tags" there is a lot of common stuff going on, isnit

Reply:
I think I understand